eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java

```diff
@@ -905,13 +905,41 @@ public T detect(Predicate<? super T> predicate)
     }
 
     @Override
+    public <P> T detectWith(Predicate2<? super T, ? super P> predicate, P parameter)
+    {
+        return InternalArrayIterate.detectWith(this.items, this.size, predicate, parameter);
+    }
+
+    @Override
+    public Optional<T> detectOptional(Predicate<? super T> predicate)
+    {
+        return InternalArrayIterate.detectOptional(this.items, this.size, predicate);
+    }
+
+    @Override
+    public <P> Optional<T> detectWithOptional(Predicate2<? super T, ? super P> predicate, P parameter)
+    {
+        return InternalArrayIterate.detectWithOptional(this.items, this.size, predicate, parameter);
+    }
+
+    @Override
     public T detectIfNone(Predicate<? super T> predicate, Function0<? extends T> defaultValueBlock)
     {
         T result = this.detect(predicate);
         return result == null ? defaultValueBlock.value() : result;
     }
 
     @Override
+    public <P> T detectWithIfNone(
+            Predicate2<? super T, ? super P> predicate,
+            P parameter,
+            Function0<? extends T> defaultValueBlock)
+    {
+        T result = this.detectWith(predicate, parameter);
+        return result == null ? defaultValueBlock.value() : result;
+    }
+
```

Comment:
If detectOptional returns a present Optional, then function should not be
invoked and passing a null function should be ok.
On Wed, Aug 31, 2016 at 12:50 PM Hiroshi Ito <notifications@github.com>
wrote:

> In
> eclipse-collections/src/main/java/org/eclipse/collections/impl/list/mutable/FastList.java
> <https://github.com/eclipse/eclipse-collections/pull/136#discussion_r77025808>
> :
>
> >      public T detectIfNone(Predicate<? super T> predicate, Function0<? extends T> defaultValueBlock)
> >      {
> >          T result = this.detect(predicate);
> >          return result == null ? defaultValueBlock.value() : result;
> >      }
> >
> >      @Override
> > +    public <P> T detectWithIfNone(
> > +            Predicate2<? super T, ? super P> predicate,
> > +            P parameter,
> > +            Function0<? extends T> defaultValueBlock)
> > +    {
> > +        T result = this.detectWith(predicate, parameter);
> > +        return result == null ? defaultValueBlock.value() : result;
> > +    }
> > +
>
> Yes, this was the code that was failing with NPE (it was in
> AbstractRichIterable not in FastList so using function instead of
> defaultValueBlock).
>
>     @Override
>     public T detectIfNone(Predicate<? super T> predicate, Function0<? extends T> function)
>     {
>         return this.detectOptional(predicate).orElseGet(function::value);
>     }
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/eclipse/eclipse-collections/pull/136/files/16e1cc60f767f4b860f287066f53699d53403ae8#r77025808>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAO6IorUWseFqfM4VYFFCQUvddMYz5drks5qlbDCgaJpZM4Jol8s>
> .
>